### DISCUSSION TITLE
Bulk adding of ORM objects in async code

### DISCUSSION THREAD

**Discussion by qarmin** [sqlalchemy/sqlalchemy#8267] [Usage Questions]
Hi, 
I have similar structure in my project which use Sqlite(at least for now) to this
```
class Course:
    __tablename__ = "course"
    id = Column(Integer, primary_key=True)
    points = relationship(Point, lazy="subquery")
    
class Point(Base):
    __tablename__ = "point"
    id = Column(Integer, primary_key=True)
    course_id = Column(Integer, ForeignKey("course.id"))
    expositions = relationship(Exposition, lazy="subquery")

class Exposition(Base):
    __tablename__ = "exposition"
    id = Column(Integer, primary_key=True)
    point_id = Column(Integer, ForeignKey("exposition.id"))
```
In one place I'm creating object in this way
```
course = Course()
for i in range(20):
     point = Point()
     for i in range(20):
           exposition = Exposition()
           point.append(exposition)
      course.append(point)
      
session.add(course)
await session.commit()
```
but it produces almost 400 inserts which on device that I use(4x1Ghz) takes usually few seconds.
It is possible to batch them to execute one(or only few) inserts?

I read before about other similar issues like https://github.com/sqlalchemy/sqlalchemy/discussions/7869, but still don't know if it even works.

**Comment by zzzeek** [comment]
until we implement #6047 for 2.0, which is literally what I hope to start this week, you need to assign the primary key to your objects up front.  then it will batch the inserts since it does not need to post-fetch the newly generated primary key value.   you can do this by bumping the sequence assigned to the integer id of the table if you are using a SERIAL value.
